features/equal/equal.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) 2022 PlanetScale Inc. All rights reserved.
+
+package equal
+
+import (
+	"fmt"
+	"sort"
+
+	"google.golang.org/protobuf/compiler/protogen"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"github.com/planetscale/vtprotobuf/generator"
+)
+
+func init() {
+	generator.RegisterFeature("equal", func(gen *generator.GeneratedFile) generator.FeatureGenerator {
+		return &equal{GeneratedFile: gen}
+	})
+}
+
+type equal struct {
+	*generator.GeneratedFile
+	once bool
+}
+
+var _ generator.FeatureGenerator = (*equal)(nil)
+
+func (p *equal) Name() string     { return "equal" }
+func (p *equal) GenerateHelpers() {}
+func (p *equal) GenerateFile(file *protogen.File) bool {
+	proto3 := file.Desc.Syntax() == protoreflect.Proto3
+	for _, message := range file.Messages {
+		p.message(proto3, message)
+	}
+	return p.once
+}
+
+const equalName = "EqualVT"
+
+func (p *equal) message(proto3 bool, message *protogen.Message) {
+	for _, nested := range message.Messages {
+		p.message(proto3, nested)
+	}
+
+	if message.Desc.IsMapEntry() {
+		return
+	}
+
+	p.once = true
+
+	ccTypeName := message.GoIdent
+	p.P(`func (this *`, ccTypeName, `) `, equalName, `(that *`, ccTypeName, `) bool {`)
+
+	p.P(`if this == nil {`)
+	p.P(`	return that == nil || `, p.Ident("fmt", "Sprintf"), `("%v", that) == ""`)
+	p.P(`} else if that == nil {`)
+	p.P(`	return `, p.Ident("fmt", "Sprintf"), `("%v", this) == ""`)
+	p.P(`}`)
+
+	sort.Slice(message.Fields, func(i, j int) bool {
+		return message.Fields[i].Desc.Number() < message.Fields[j].Desc.Number()
+	})
+
+	{
+		oneofs := make(map[string]struct{}, len(message.Fields))
+		scoped := false
+		for _, field := range message.Fields {
+			oneof := field.Oneof != nil && !field.Oneof.Desc.IsSynthetic()
+			nullable := field.Message != nil || (field.Oneof != nil && field.Oneof.Desc.IsSynthetic()) || (!proto3 && !oneof)
+			if oneof {
+				fieldname := field.Oneof.GoName
+				if _, ok := oneofs[fieldname]; !ok {
+					oneofs[fieldname] = struct{}{}
+					p.P(`if this.`, fieldname, ` == nil && that.`, fieldname, ` != nil {`)
+					p.P(`	return false`)
+					p.P(`} else if this.`, fieldname, ` != nil {`)
+					p.P(`	if that.`, fieldname, ` == nil {`)
+					p.P(`		return false`)
+					p.P(`	}`)
+					scoped = true
+				}
+
+				p.oneof(field, nullable)
+			}
+		}
+		if scoped {
+			p.P(`}`)
+		}
+	}
+
+	for _, field := range message.Fields {
+		oneof := field.Oneof != nil && !field.Oneof.Desc.IsSynthetic()
+		nullable := field.Message != nil || (field.Oneof != nil && field.Oneof.Desc.IsSynthetic()) || (!proto3 && !oneof)
+		if !oneof {
+			p.field(field, nullable)
+		}
+	}
+
+	p.P(`return string(this.unknownFields) == string(that.unknownFields)`)
```

Comment:
This should be `bytes.Equal`. It doesn't make much sense to force a `string` cast for comparison.